### PR TITLE
niza/initial-setup

### PR DESCRIPTION
**Pull request by nizacegodk**
Added initial abstract service client implementation

### CODE REVIEW

File: src/ServiceClientBase/Exceptions/ServiceRequestFailedException.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+use Throwable;
+use Illuminate\Http\Client\Response;
+
+/**
+ * Class SeamlessWalletRequestFailedException
+ */
+class ServiceRequestFailedException extends Exception
+{
+    /**
+     * SeamlessWalletRequestFailedException constructor.
```
Comment (wogelius) [inline src/ServiceClientBase/Exceptions/ServiceRequestFailedException.php]: (minus) SeamlessWallet

File: src/ServiceClientBase/Exceptions/ServiceRequestFailedException.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+     *
+     * @param Response $response
+     * @param Throwable|null $previous
+     */
+    public function __construct(Response $response, Throwable $previous = null)
+    {
+        $message = sprintf("Seamless Wallet Service [%s]: \n %s", $response->status(), $response->body());
```
Comment (wogelius) [inline src/ServiceClientBase/Exceptions/ServiceRequestFailedException.php:22]: (minus) Seamless Wallet

File: src/ServiceClientBase/RequestDrivers/HttpRequestDriver.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+     * @param array $headers
+     * @param array $options
+     *
+     * @return Response
+     *
+     * @throws ServiceRequestFailedException
+     */
+    public function makeRequest(string $method, string $endpoint, array $data = [], array $headers = [], array $options = []): Response
+    {
+        $maxTries = env("SEAMLESS_WALLET_CLIENT_MAXIMUM_NUMBER_OF_RETRIES", 3);
```
Comment (wogelius) [inline src/ServiceClientBase/RequestDrivers/HttpRequestDriver.php:26]: Generelt omdøb SEAMLESS_WALLET til SERVICE_CLIENT i denne fil

**Review by wogelius — COMMENTED** [review]
Just a couple of small comments

File: src/ServiceClientBase/AbstractServiceClient.php [hunk 4]
```diff
@@ -1,6 +1,287 @@
...
+            throw new InvalidHeaderException('Header name and value must be strings!');
+        }
+
+        $this->globalHeaders[$header] = $value;
+
+        return $this;
+    }
+
+    /**
+     * Push headers to the global headers array, which applies these headers to all requests made by the client.
```
[the remark this answers is not in the document]
Comment (nizacegodk) [reply src/ServiceClientBase/AbstractServiceClient.php]: Fixed

File: src/ServiceClientBase/Exceptions/ServiceRequestFailedException.php (hunk 2 above)
Comment (nizacegodk) [reply src/ServiceClientBase/Exceptions/ServiceRequestFailedException.php:22]: Fixed

File: src/ServiceClientBase/Exceptions/ServiceRequestFailedException.php (hunk 1 above)
Comment (nizacegodk) [reply src/ServiceClientBase/Exceptions/ServiceRequestFailedException.php]: Fixed

File: src/ServiceClientBase/RequestDrivers/HttpRequestDriver.php (hunk 3 above)
Comment (nizacegodk) [reply src/ServiceClientBase/RequestDrivers/HttpRequestDriver.php]: Fixed

**Pull request merged**